File: backends/dotnet/services/Geo/Geo.Client/GeoClientOptions.cs
// -----------------------------------------------------------------------
// <copyright file="GeoClientOptions.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Client;

/// <summary>
/// Configuration options for the Geo Client library.
/// </summary>
public class GeoClientOptions
{
    /// <summary>
    /// Gets or sets the duration after which WhoIs cache entries expire.
    /// </summary>
    public TimeSpan WhoIsCacheExpiration { get; set; } = TimeSpan.FromHours(8);

    /// <summary>
    /// Gets or sets the maximum number of WhoIs entries in the local cache (LRU eviction).
    /// </summary>
    public int WhoIsCacheMaxEntries { get; set; } = 10_000;

    /// <summary>
    /// Gets or sets the context keys allowed for contact operations.
    /// Client-side defense-in-depth validation. Empty list disables validation.
    /// </summary>
    public List<string> AllowedContextKeys { get; set; } = [];

    /// <summary>
    /// Gets or sets the API key sent via gRPC metadata (x-api-key header) for Geo service authentication.
    /// </summary>
    public string ApiKey { get; set; } = string.Empty;
}

File: backends/node/services/geo/geo-client/src/geo-client-options.ts
/** Configuration options for the Geo Client library. */
export interface GeoClientOptions {
  /** Duration in ms after which WhoIs cache entries expire. Default: 28,800,000 (8 hours). */
  whoIsCacheExpirationMs: number;
  /** Maximum number of WhoIs entries in the local cache (LRU eviction). Default: 10,000. */
  whoIsCacheMaxEntries: number;
  /** Directory for persisted data files. Default: "./data". */
  dataDir: string;
  /** Maximum number of Contact entries in the local cache (LRU eviction). Default: 10,000. No TTL — contacts are immutable. */
  contactCacheMaxEntries: number;
  /** Client-side validation: only these context keys are allowed for contact operations. Empty array disables validation. */
  allowedContextKeys: string[];
  /** API key sent via gRPC metadata (`x-api-key` header) for Geo service authentication. */
  apiKey: string;
  /** Timeout in ms for gRPC calls to the Geo service. Default: 30,000 (30 seconds). */
  grpcTimeoutMs: number;
  /** Duration in ms for negative cache entries (WhoIs not found). Default: 3,600,000 (1 hour). */
  whoIsNegativeCacheExpirationMs: number;
  /** Consecutive gRPC failures before the circuit breaker opens. Default: 5. */
  circuitBreakerFailureThreshold: number;
  /** Duration in ms the circuit stays open before probing. Default: 30,000 (30s). */
  circuitBreakerCooldownMs: number;
}

export const DEFAULT_GEO_CLIENT_OPTIONS: GeoClientOptions = {
  whoIsCacheExpirationMs: 28_800_000,
  whoIsCacheMaxEntries: 10_000,
  dataDir: "./data",
  contactCacheMaxEntries: 10_000,
  allowedContextKeys: [],
  apiKey: "",
  grpcTimeoutMs: 30_000,
  whoIsNegativeCacheExpirationMs: 3_600_000,
  circuitBreakerFailureThreshold: 5,
  circuitBreakerCooldownMs: 30_000,
};
